src/render/draw_heatmap.js

```diff
@@ -0,0 +1,137 @@
+// @flow
+
+const browser = require('../util/browser');
+const EXTENT = require('../data/extent');
+const mat4 = require('@mapbox/gl-matrix').mat4;
+const Texture = require('./texture');
+
+import type Painter from './painter';
+import type SourceCache from '../source/source_cache';
+import type HeatmapStyleLayer from '../style/style_layer/heatmap_style_layer';
+import type HeatmapBucket from '../data/bucket/heatmap_bucket';
+import type TileCoord from '../source/tile_coord';
+
+module.exports = drawHeatmap;
+
+function drawHeatmap(painter: Painter, sourceCache: SourceCache, layer: HeatmapStyleLayer, coords: Array<TileCoord>) {
+    if (painter.isOpaquePass) return;
+
+    const gl = painter.gl;
+
+    painter.setDepthSublayer(0);
+    painter.depthMask(false);
+
+    // Allow circles to be drawn across boundaries, so that
+    // large circles are not clipped to tiles
+    gl.disable(gl.STENCIL_TEST);
+
+    renderToTexture(gl, painter, layer);
+
+    gl.clearColor(0, 0, 0, 0);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    gl.blendFunc(gl.ONE, gl.ONE);
+
+    for (let i = 0; i < coords.length; i++) {
+        const coord = coords[i];
+
+        // skip tiles that have uncovered parents to avoid flickering
```

Comment:
Can you please expand on the reasoning of this a little more? And potentially add a render test?

Reply:
Certainly needs a render test, although it will be tricky to make because of the need to manually make 4 new tile fixtures. The situation here is similar to #5004. The difference is that heatmaps don't need such a complex fix (tile masks) — the difference between tiles of different zooms for heatmaps is much more subtle since it's still vector data, so waiting for all 4 children to load before putting them in place of a parent is an acceptable solution to the problem that's also faster and much simpler.